Dockerfile

```diff
@@ -1,20 +1,32 @@
-FROM crystallang/crystal:1.2.2-alpine AS builder
+FROM docker.io/node:18-alpine AS node-builder
+RUN mkdir /build
+WORKDIR /build
+COPY --chown=0:0 package.json package-lock.json /build/
+RUN npm ci
+COPY --chown=0:0 webpack.config.js ./
+RUN mkdir src
+COPY --chown=0:0 src/assets src/assets
+RUN npm run build
+
+FROM docker.io/crystallang/crystal:1.7.2-alpine AS crystal-builder
 RUN apk update && apk upgrade && apk add sqlite-static
 WORKDIR /build/
 ARG version
-RUN git clone --branch ${version:-dist} --depth 1 https://github.com/toddsundsted/ktistec .
+COPY --chown=0:0 shard.yml shard.lock ./
 RUN shards install --production
-RUN crystal build src/ktistec/server.cr --static --no-debug --release
+COPY --chown=0:0 etc etc
+COPY --chown=0:0 src src
+RUN crystal build src/ktistec/server.cr --progress --static --no-debug --release
```

Comment:
This is the same logic as the node build, but to produce the crystal binary.

We require the repo to contain the share.yml and shard.lock to make the build reproducible. Then only re-install the dependencies if those two files changed.

I added `--progress` to the build because I found that the crystal compiled doesn't provide much feedback and generally takes quite a bit of time to provide an output.